### PR TITLE
Limit precision to 999,999,999

### LINKED ISSUE #30352: [Breaking change]: Limiting the precision in numeric format strings from Int32.MaxValue to 999,999,999
### Description

The fix will likely be merged into .NET 7 RC1 with dotnet/runtime#72647.

in .NET 6, we updated the maximum precision for numeric format strings to be `Int32.MaxValue`, as documented here https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings and here https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/numeric-format-parsing-handles-higher-precision. Unfortunately, the implementation was buggy, and the easiest fix requires we pick a slightly smaller limit of `999,999,999`. The original expansion to `Int32.MaxValue` was pretty arbitrary as far as I can tell, and `999,999,999` is more than big enough. This comment (https://github.com/dotnet/runtime/issues/67891#issuecomment-1190783778) goes into a bit more detail on the logic of this choice.

### Version

.NET 7 RC1

### Previous behavior

The previous behavior *intended* to throw a `FormatException` for any precision larger than `Int.MaxValue`. It did not throw that exception for many of such inputs due to the mentioned bug. The intended behavior was:
```c#
double d = 123.0;

d.ToString("E" + int.MaxValue.ToString()); // doesn't throw

long intMaxPlus1 = (long)int.MaxValue + 1;
string intMaxPlus1String = intMaxPlus1.ToString();
Assert.Throws<FormatException>(() => d.ToString("E" + intMaxPlus1String)); // throws
```

### New behavior

Now, we correctly throw a `FormatException` for any precision larger than `999,999,999`.

```c#
double d = 123.0;
Assert.Throws<FormatException>(() => d.ToString("E" + int.MaxValue.ToString())); // throws

long intMaxPlus1 = (long)int.MaxValue + 1;
string intMaxPlus1String = intMaxPlus1.ToString();
Assert.Throws<FormatException>(() => d.ToString("E" + intMaxPlus1String)); // throws

d.ToString("E999999999"); // doesn't throw

d.ToString("E00000999999999"); // doesn't throw
```

### Type of breaking change

- [x] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load/execute or different run-time behavior.
- [ ] **Source incompatible**: Source code may encounter a breaking change in behavior when targeting the new runtime/component/SDK, such as compile errors or different run-time behavior.

### Reason for change

Described above.

### Recommended action

This change should not have any substantial effect, as I don't think many are using precisions this large. Simply updating the documentation should be sufficient.

### Feature area

Core .NET libraries

### Affected APIs

Should be the same APIs as the previous change https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/numeric-format-parsing-handles-higher-precision

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #30352.

[Internal preview](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings?branch=pr-en-us-31002).

### CODE REVIEW

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+.NET 7 RC 1
+
+## Type of breaking change
+
+This change can affect [binary compatibility](../../categories.md#binary-compatibility).
+
+## Reason for change
+
+The behavior that was introduced in .NET 6 was intended to throw a <xref:System.FormatException> for any precision larger than <xref:System.Int32.MaxValue?displayProperty=nameWithType>. However, due to a bug, it did not throw that exception for some inputs larger than the maximum. This change fixes the bug by limiting the precision to 999,999,999.
```
Comment (adegeo) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:57]: I don't like this phrase:

> This change fixes the bug by limiting the precision to 999,999,999.

The fix isn't the limit being set, the limit was in the previous version. The fix is the exception is thrown.

**Review by adegeo — APPROVED** [review]

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
+---
+title: "Breaking change: Maximum precision for numeric format strings"
+description: Learn about the .NET 7 breaking change in core .NET libraries where the maximum precision for numeric format strings was changed to 999,999,999.
+ms.date: 09/02/2022
+---
+# Maximum precision for numeric format strings
+
+The maximum precision when formatting numbers as strings using `ToString` and `TryFormat` has been changed from <xref:System.Int32.MaxValue?displayProperty=nameWithType> to 999,999,999. (The maximum precision was [previously changed](../6.0/numeric-format-parsing-handles-higher-precision.md) to <xref:System.Int32.MaxValue?displayProperty=nameWithType> in .NET 6.)
+
+In addition, the maximum exponent allowed when parsing an <xref:System.Int64> from a string has been limited to 999,999,999.
```
Comment (dakersnar) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:10]:
```suggestion
In addition, the maximum exponent allowed when parsing a <xref:System.Numerics.BigInteger> from a string has been limited to 999,999,999.
```

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+double d = 123.0;
+
+d.ToString("E" + int.MaxValue.ToString()); // Doesn't throw.
+
+long intMaxPlus1 = (long)int.MaxValue + 1;
+string intMaxPlus1String = intMaxPlus1.ToString();
+Assert.Throws<FormatException>(() => d.ToString("E" + intMaxPlus1String)); // Throws.
+```
+
+In addition, there was no limit on the exponent size when parsing an <xref:System.Int64> from a string.
```
Comment (dakersnar) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:26]:
```suggestion
In addition, there was no limit on the exponent size when parsing a <xref:System.Numerics.BigInteger> from a string.
```

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+long intMaxPlus1 = (long)int.MaxValue + 1;
+string intMaxPlus1String = intMaxPlus1.ToString();
+Assert.Throws<FormatException>(() => d.ToString("E" + intMaxPlus1String)); // Throws.
+
+d.ToString("E999999999"); // Doesn't throw.
+
+d.ToString("E00000999999999"); // Doesn't throw.
+```
+
+In addition, if you attempt to parse an <xref:System.Int64> with an exponent greater than 999,999,999 from a string, a <xref:System.FormatException> is thrown.
```
Comment (dakersnar) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:45]:
```suggestion
In addition, if you attempt to parse a <xref:System.Numerics.BigInteger> with an exponent greater than 999,999,999 from a string, a <xref:System.FormatException> is thrown.
```

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+- <xref:System.SByte.ToString(System.String,System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.SByte.TryFormat(System.Span{System.Char},System.Int32@,System.ReadOnlySpan{System.Char},System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.Int16.ToString(System.String)?displayProperty=fullName>
+- <xref:System.Int16.ToString(System.String,System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.Int16.TryFormat(System.Span{System.Char},System.Int32@,System.ReadOnlySpan{System.Char},System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.UInt16.ToString(System.String)?displayProperty=fullName>
+- <xref:System.UInt16.ToString(System.String,System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.UInt16.TryFormat(System.Span{System.Char},System.Int32@,System.ReadOnlySpan{System.Char},System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.Int64.Parse%2A?displayProperty=fullName>
+- <xref:System.Int64.TryParse%2A?displayProperty=fullName>
```
Comment (dakersnar) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:89]: I think these have to be updated to BigInteger if my other corrections are accurate.

**Review by dakersnar — COMMENTED** [review]
LGTM, but I think the references to parsing `Int64` with an exponent should be in reference to `BigInteger`, if this is the change you are referring to https://github.com/dotnet/runtime/pull/73643

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+- <xref:System.SByte.ToString(System.String)?displayProperty=fullName>
+- <xref:System.SByte.ToString(System.String,System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.SByte.TryFormat(System.Span{System.Char},System.Int32@,System.ReadOnlySpan{System.Char},System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.Int16.ToString(System.String)?displayProperty=fullName>
+- <xref:System.Int16.ToString(System.String,System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.Int16.TryFormat(System.Span{System.Char},System.Int32@,System.ReadOnlySpan{System.Char},System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.UInt16.ToString(System.String)?displayProperty=fullName>
+- <xref:System.UInt16.ToString(System.String,System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.UInt16.TryFormat(System.Span{System.Char},System.Int32@,System.ReadOnlySpan{System.Char},System.IFormatProvider)?displayProperty=fullName>
+- <xref:System.Int64.Parse%2A?displayProperty=fullName>
```
Comment (gewarren) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:88]:
```suggestion
- <xref:System.Numerics.BigInteger.Parse%2A?displayProperty=fullName>
```

File: docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md (hunk 5 above)
Comment (gewarren) [inline docs/core/compatibility/core-libraries/7.0/max-precision-numeric-format-strings.md:89]:
```suggestion
- <xref:System.Numerics.BigInteger.TryParse%2A?displayProperty=fullName>
```

**Pull request merged**